### PR TITLE
add steps to write modulewithloss wrapper

### PR DESCRIPTION
**Pull request by AdamLouly**
### Description
This PR includes documentation updates, providing step-by-step instructions on how to implement the ModuleWithLoss wrapper in a different codebase. 
The documentation outlines the necessary code changes and offers customization options based on specific requirements.

### CODE REVIEW

File: docs/ORTModule_ModuleWithLoss_Wrapper.md [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+
+#### ModuleWithLoss Wrapper
+
+- **Description**: The ModuleWithLoss wrapper is a class that extends the nn.Module class from PyTorch. It is designed to enhance the functionality of an existing model by allowing the computation of loss inside the wrapper, leveraging the benefits of ONNX Runtime (ORT) and enabling additional optimizations such as label sparsity optimization.
+
+These are the Major changes on the code, Refer to these snippets as a reference while writing your own implementation.
+
+```python
+class ModuleWithLoss(nn.Module):
+    def __init__(self, model, args, label_smoother):
```
Comment (prathikr) [inline docs/ORTModule_ModuleWithLoss_Wrapper.md:13]: Can you add types for all input parameters like you did for `inputs: Dict[...]` on line 20?

File: docs/ORTModule_ModuleWithLoss_Wrapper.md [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+        super().__init__()
+        self._original_model = model
+        self.args = args
+        # Label smoothing
+        self.label_smoother = label_smoother
+
+    def forward(self, inputs: Dict[str, Union[torch.Tensor, Any]], return_outputs):
```
Comment (prathikr) [inline docs/ORTModule_ModuleWithLoss_Wrapper.md:20]: same comment as line 13

File: docs/ORTModule_ModuleWithLoss_Wrapper.md [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+
+## Initializing the Wrapper
+
+### 4. Create a function to initialize the wrapper:
+- Implement a function, such as `create_model_with_loss()`, to create an instance of `ModuleWithLoss` using the original model, arguments, and label smoother.
+- Assign `compute_model_plus_loss_internal` as a method of `Trainer.compute_loss` to the wrapper instance.
+- Return the created wrapper instance.
+
+### 5. Override the necessary functions:
+- Override the `_set_signature_columns_if_needed` function if needed.
```
Comment (prathikr) [inline docs/ORTModule_ModuleWithLoss_Wrapper.md:96]: Can you give some more details on what this function does and when it is necessary to override vs. when its ok to leave it?

File: docs/ORTModule_ModuleWithLoss_Wrapper.md [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+        if isinstance(self.model, ModuleWithLoss):
+            signature = inspect.signature(self.model._original_model.forward)
+        else:
+            signature = inspect.signature(self.model.forward)
+
+        self._signature_columns = list(signature.parameters.keys())
+        # Labels may be named label or label_ids, the default data collator handles that.
+        self._signature_columns += list(set(["label", "label_ids"] + self.label_names))
+
+def compute_loss(self, model_with_loss, inputs, return_outputs=False):
```
Comment (prathikr) [inline docs/ORTModule_ModuleWithLoss_Wrapper.md:57]: Can we instead override compute_loss inside the ModuleWithLoss base class?

File: docs/ORTModule_ModuleWithLoss_Wrapper.md [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+    def config(self):
+        return self._original_model.config
+
+def create_model_with_loss(self):
+    model_with_loss = ModuleWithLoss(self.model, self.args, self.label_smoother)
+    model_with_loss.compute_model_plus_loss_internal = types.MethodType(Trainer.compute_loss, model_with_loss)
+
+    return model_with_loss
+
+def _set_signature_columns_if_needed(self):
```
Comment (prathikr) [inline docs/ORTModule_ModuleWithLoss_Wrapper.md:43]: Can we instead override `_set_signature_columns_if_needed(...)` inside the ModuleWithLoss base class?

File: docs/ORTModule_ModuleWithLoss_Wrapper.md (hunk 1 above)
Comment (AdamLouly) [reply docs/ORTModule_ModuleWithLoss_Wrapper.md:13]: This code is taken from the optimum trainer.py implementation, so just copied and pasting it here in this doc so people can see the implementation, changes on the code must be on the optimum repo.
Will keep these change requests in consideration when modifying the optimum code next time.

File: docs/ORTModule_ModuleWithLoss_Wrapper.md (hunk 4 above)
Comment (AdamLouly) [reply docs/ORTModule_ModuleWithLoss_Wrapper.md:57]: For the current implementation, we can compute loss by wrapping it only in ModuleWithLoss because it is used in both ORTTrainer and ModuleWithLoss. Additionally, we are still utilizing the flag --use_module_with_loss. 
once this behavior turn to be default, you have the flexibility to define your code wherever you prefer.

File: docs/ORTModule_ModuleWithLoss_Wrapper.md (hunk 5 above)
Comment (AdamLouly) [reply docs/ORTModule_ModuleWithLoss_Wrapper.md:43]: same as previous

File: docs/ORTModule_ModuleWithLoss_Wrapper.md [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+        self._original_model = model
+        self.args = args
+        # Label smoothing
+        self.label_smoother = label_smoother
+
+    def forward(self, inputs: Dict[str, Union[torch.Tensor, Any]], return_outputs):
+        # The compute_model_plus_loss_internal is assigned once the class is instantiated.
+        # It should have same signature as Trainer.compute_loss().
+        # We do this to avoid potential un-synced states if we duplicated compute loss codes .
+        return self.compute_model_plus_loss_internal(self._original_model, inputs, return_outputs)
```
Comment (pengwa) [inline docs/ORTModule_ModuleWithLoss_Wrapper.md:24]: this code is very specialized for optimum trainer, I would suggest not to just copy that as a common example. 

You can define a nn.Module which only compute stacks of layers, and return the outputs. the training script run the nn.Module's forward, and then feed its outputs into cross entropy loss, then trigger a backward. 

This is more like an example we want to cover using this doc.

File: docs/ORTModule_ModuleWithLoss_Wrapper.md [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+### 4. Create a function to initialize the wrapper:
+- Implement a function, such as `create_model_with_loss()`, to create an instance of `ModuleWithLoss` using the original model, arguments, and label smoother.
+- Assign `compute_model_plus_loss_internal` as a method of `Trainer.compute_loss` to the wrapper instance.
+- Return the created wrapper instance.
+
+### 5. Override the necessary functions:
+- Override the `_set_signature_columns_if_needed` function if needed.
+- Update the function to handle the case where the model is an instance of `ModuleWithLoss`.
+
+### 6. Override the `compute_loss` function if needed:
```
Comment (pengwa) [inline docs/ORTModule_ModuleWithLoss_Wrapper.md:99]: the steps are specialized, I don't think for other training script, users can follow those steps  for warpping the loss.

**Review by prathikr — APPROVED** [review]

File: docs/ORTModule_ModuleWithLoss_Wrapper.md [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+1. Create an instance of `ModuleWithLoss` by passing the original model as an argument.
+2. Initialize the optimizer for training.
+3. Enter the training loop and iterate over the training data.
+4. Zero the gradients of the optimizer.
+5. Compute the loss by calling the `compute_loss` function and passing the model, inputs, and targets.
+6. Perform the backward pass and optimization step by calling `loss.backward()` and `optimizer.step()`.
+
+
+## Full Example
+
```
Comment (pengwa) [inline docs/ORTModule_ModuleWithLoss_Wrapper.md:32]:
```diff

    # Define the model architecture
    class PretrainedModel(nn.Module):
        ...

        def forward(self, input_ids, attention_mask):
            ...
            transformer_outputs = self.transformer(
                input_ids,
                attention_mask=attention_mask,
                ...
            )
            hidden_states = transformer_outputs[0]
            lm_logits = self.lm_head(hidden_states)
            return lm_logits

+    class ModuleWithLoss(nn.Module):
+        def __init__(self, model):
+            super().__init__()
+            self.model = model

+        def forward(self, *inputs):
+            # Perform the forward pass of the model and return the outputs
+            outputs = self.model(inputs)

+            # Compute the forward pass of the model
+            lm_logits= model(inputs)

+            # Compute the cross-entropy loss
+            loss = nn.CrossEntropyLoss()(lm_logits, labels)
+            return loss


    # Training loop
    model = PretrainedModel(...)
+    model = ModuleWithLoss(model)

    optimizer = torch.optim.Adam(model.parameters(), lr=0.001)

    model = ORTModule(model)

    for inputs, labels in dataloader:
        optimizer.zero_grad()

-        # Compute the forward pass of the model
-        lm_logits= model(inputs)

-        # Compute the cross-entropy loss
-        loss = nn.CrossEntropyLoss()(outputs, labels)


+        # Compute the forward pass and cross-entropy loss
+        loss= model(inputs)

        # Backward pass and optimization step
        loss.backward()
        optimizer.step()

```

Comment (pengwa) [reply docs/ORTModule_ModuleWithLoss_Wrapper.md:32]: I feel the example script is not explaining loss wrapping correctly. I put an example above, which is a diff view, comparing origional model train script with the modified model train script. Does that looks better to be self-explained?

Comment (pengwa) [reply docs/ORTModule_ModuleWithLoss_Wrapper.md:32]: The steps need to be adapted according to it.

Comment (AdamLouly) [reply docs/ORTModule_ModuleWithLoss_Wrapper.md:32]: Thanks for the suggestion, will modify it accordingly.
Thanks

File: docs/ORTModule_ModuleWithLoss_Wrapper.md [hunk 9]
```diff
@@ -0,0 +1,94 @@
+# ONNX Runtime ModuleWithLoss Wrapper
+
+This document provides instructions on implementing a wrapper similar to the ModuleWithLoss Wrapper in Optimum. By implementing this wrapper, you can compute the loss inside ONNX Runtime (ORT), enabling you to leverage additional optimizations such as label sparsity optimization.
+## Implementation Steps
+
+Follow these steps to create your own ModuleWithLoss for computing loss inside ONNX Runtime:
+
+Certainly! Here are the steps to fit the provided example:
+
+### Step 1: Define `PretrainedModel` Class
```
Comment (pengwa) [inline docs/ORTModule_ModuleWithLoss_Wrapper.md:10]: step1 and step 3 are the model user has by default, maybe we can just have a section "Define a sample training script" which has a PretrainedModule , and the training loops. 

Another section is to explain the steps to wrap the loss compute in ModuleWIthLoss, and adaption in the training loops.

File: docs/ORTModule_ModuleWithLoss_Wrapper.md [hunk 10]
```diff
@@ -0,0 +1,94 @@
+# ONNX Runtime ModuleWithLoss Wrapper
+
+This document provides instructions on implementing a wrapper similar to the ModuleWithLoss Wrapper in Optimum. By implementing this wrapper, you can compute the loss inside ONNX Runtime (ORT), enabling you to leverage additional optimizations such as label sparsity optimization.
```
Comment (pengwa) [inline docs/ORTModule_ModuleWithLoss_Wrapper.md:3]: the daption is not needed for all cases, shall we clarify in which case we need it? e.g. when 1). loss is not computed in the model forward path, 2). or model forward compute the loss but also return other outputs that are not needed for subsquent computation.

The impact of 1) is ORT cannot track loss compute in ONNX graph; impact of 2) is  ORT graph return more tensors that is needed, which will take unnecessary CUDA memory during backwards.

**Review by pengwa — COMMENTED** [review]
Sorry for be picky, I feel there are somethings we need clarify to users for the intention, the cases where it is needed, and a clear diff changed based on user's original pytorch model (in this case it is the sample script).

File: docs/ORTModule_ModuleWithLoss_Wrapper.md (hunk 9 above)
Comment (AdamLouly) [reply docs/ORTModule_ModuleWithLoss_Wrapper.md]: I put the 2 steps under one step, and added the code to each one, so it can be more practical.

File: docs/ORTModule_ModuleWithLoss_Wrapper.md (hunk 10 above)
Comment (AdamLouly) [reply docs/ORTModule_ModuleWithLoss_Wrapper.md:3]: Yeah I think giving context to why we're doing this and its impact will be useful.
Thanks, Will add this

**Review by pengwa — APPROVED** [review]

File: docs/ORTModule_ModuleWithLoss_Wrapper.md (hunk 10 above)
Comment (pengwa) [reply docs/ORTModule_ModuleWithLoss_Wrapper.md:3]: Thanks for making this change!
